client/client.go

```diff
@@ -401,27 +410,51 @@ func (c *httpClusterClient) Sync(ctx context.Context) error {
 		return err
 	}
 
-	c.Lock()
-	defer c.Unlock()
-
 	var eps []string
 	for _, m := range ms {
 		eps = append(eps, m.ClientURLs...)
 	}
-	sort.Sort(sort.StringSlice(eps))
 
-	ceps := make([]string, len(c.endpoints))
-	for i, cep := range c.endpoints {
-		ceps[i] = cep.String()
+	neps, err := c.parseEndpoints(eps)
+	if err != nil {
+		return err
 	}
-	sort.Sort(sort.StringSlice(ceps))
-	// fast path if no change happens
-	// this helps client to pin the endpoint when no cluster change
-	if reflect.DeepEqual(eps, ceps) {
-		return nil
+
+	npin := 0
+
+	switch c.selectionMode {
+	case EndpointSelectionRandom:
```

Comment:
There's actually a test for the pinning behaviour here, `TestHTTPClusterClientSyncPinEndpoint`. That would only persist the pin when the member list was identical to the previous list, which is slightly different to this (which persists the pin as long as the pinned member is in the new member list). 

For a brand new client, the pin will still be zero, but the endpoint list will have been shuffled when `New` calls `SetEndpoints`. We'll persist that pin here, so the pin shouldn't stay zero. Not exactly clear though.